README.md
# AElf Block Scan

A tool to scan the AElf Chain.
* You can use this library to scan the specific AElf node and get the blocks and transactions in blocks.
* By implementing the `DBBaseOperation` class, you can get the data in `insert` hook and implement the SQL operations with anything you want.

## Installation

```bash
# use npm
npm i aelf-block-scan --save

# use yarn
yarn
```

## Usage

Check the `example/index.js` as a reference

### Step.1

Create an AElf instance
```javascript
const AElf = require('aelf-sdk');
const aelf = new AElf(new AElf.providers.HttpProvider('http://18.162.41.20:8000'));
```

### Step.2

Implement the `DBBaseOperation` class.

There are three phases while scanning, and all phases can be identified by the `type` filed in `insert` function argument `data`
* Phase.1: scanning the `missingHeights` given by the `config`;
* Phase.2: scanning blocks and transactions from `startHeight` to `LIBHeight`;
* Phase.3: scanning in loop, from the last `LIBHeight` got from previous loop to current `bestHeight` in current loop.

In all phases, query blocks and transactions with a maximal concurrent limit.

There are three methods that must be implemented in sub-class of `DBBaseOperation`

* `init`: will be called before scanning, this method could include database initializing
* `insert`: called when the length of blocks already queried is equal to the `maxInsert` given in `options`, the argument `data`
is an object and contains three fields:
    * blocks: the array of blocks
    * txs: the array of transactions array, each element in `txs` is correspond to the block in `blocks` with the same `index`
    * LIBHeight: the current LIBHeight of chain (unavailable in `Phase.1`)
    * bestHeight: the current bestHeight of chain (unavailable in `Phase.1`)
* `destory`: called when error happening or exiting from the scanning process, you can close the database connection

***Notice***:
1. Notice that we query all the blocks and transactions inside blocks, and it increases the pressure of chain node.
    However we don't need all blocks and transactions for most of time, so we provide a new scan mode called `listener`
    which can get the blocks and transactions you are interested in by bloom filter. With bloom filter, you can judge whether an
    event is in a block or transaction, with this, `aelf-block-scan` can give only what you are interested in.
2. And how to use this, you only need to set the config `scanMode` to the value of `listener` and create a new array `listeners` in
   the config. Check the example [listener.js](./example/listeners.js) for more detail.



```javascript
const {
  Scanner,
  DBBaseOperation,
  QUERY_TYPE
} = require('aelf-block-scan');

class DBOperation extends DBBaseOperation {
  constructor(config) {
    super(config);
    this.lastTime = new Date().getTime();
  }

  /**
   * init before starting scanning
   */
  init() {
    console.log('init');
  }

  /**
   * will be called in every parallel scanning, could be defined as an async functions.
   * each element in `txs` is an array of transactions
   * each element in `txs` is correspond to the block in `blocks` with the same `index`
   * @param {{blocks: [], txs: [], LIBHeight: Number, bestHeight: Number}} data
   */
  async insert(data) {
    const now = new Date().getTime();
    console.log(`take time ${now - this.lastTime}ms`);
    this.lastTime = now;
    const {
      blocks,
      txs,
      type,
      bestHeight,
      LIBHeight
    } = data;
    // identify the phase
    switch (type) {
      case QUERY_TYPE.INIT:
        console.log('INIT');
        break;
      case QUERY_TYPE.MISSING:
        // there is no LIBHeight and bestHeight in data when querying missing heights
        console.log('MISSING');
        break;
      case QUERY_TYPE.GAP:
        console.log('GAP');
        console.log('LIBHeight', LIBHeight);
        break;
      case QUERY_TYPE.LOOP:
        console.log('LOOP');
        console.log('bestHeight', bestHeight);
        console.log('LIBHeight', LIBHeight);
        break;
      case QUERY_TYPE.ERROR:
        console.log('ERROR');
        break;
      default:
        break;
    }
    console.log('blocks length', blocks.length);
    console.log('transactions length', txs.reduce((acc, i) => acc.concat(i), []).length);
    if (blocks.length > 0) {
      console.log('highest height', blocks[blocks.length - 1].Header.Height);
    }
    console.log('\n\n\n');
  }

  /**
   * close sql connection or something
   */
  destroy() {
    console.log('destroy');
  }
}
```

### Step.3

Create an instance of Scanner and call `start` method

`const scanner = new Scanner(new DBOperation(), options)`;

the `options` argument is an object, the default options and valid fields are shown below:

```javascript
const defaultOptions = {
  // in Phase.3, the timeout of each loop
  // when chase up to last irreversible block height, start query in loop with this interval
  interval: 4000, // ms
  // the size of transactions per query in `aelf.chain.getTxResults`
  txResultsPageSize: 100,
  // max queries in parallel
  concurrentQueryLimit: 40,
  // query from this height, include this height
  startHeight: 0,
  // If the differences between startHeight and last irreversible block height
  // is lower than heightGap, start loop just now
  // heightGap: 50,
  // missing height list, the heights you want to query and insert
  missingHeightList: [],
  // the instance of aelf-sdk
  aelfInstance: new AElf(new AElf.providers.HttpProvider('http:127.0.0.1:8000/')),
  // max inserted Data into database
  maxInsert: 200,
  // unconfirmed block buffer
  unconfirmedBlockBuffer: 60,
  // config for log4js
  log4Config: {
    appenders: {
      transaction: {
        type: 'file',
        filename: 'transaction.log'
      },
      block: {
        type: 'file',
        filename: 'block.log'
      },
      error: {
        type: 'file',
        filename: 'error.log'
      }
    },
    categories: {
      default: { appenders: ['transaction', 'block', 'error'], level: 'info' },
      transaction: { appenders: ['transaction'], level: 'info' },
      block: { appenders: ['block'], level: 'info' }
    }
  }
}
```

Example:

```javascript
const {
  Scanner,
  DBBaseOperation,
  QUERY_TYPE
} = require('aelf-sdk');

const scanner = new Scanner(new DBOperation(), {
  aelfInstance: aelf,
  startHeight: 11000,
  missingHeightList: [12, 1414],
  interval: 8000,
  concurrentQueryLimit: 30,
  maxInsert: 100
});
scanner.start().then(res => {
  // here we just start Phase.3 loop
  console.log(res);
}).catch(err => {
  console.log(err);
});
```



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.4.1](https://github.com/AElfProject/aelf-block-scan/compare/v0.4.0...v0.4.1) (2020-03-09)


### Bug Fixes

* 🐛 fix reference ([ecb2ba3](https://github.com/AElfProject/aelf-block-scan/commit/ecb2ba3))



## [0.4.0](https://github.com/AElfProject/aelf-block-scan/compare/v0.3.1...v0.4.0) (2020-03-05)


### Features

* 🎸 add new scan mode listener ([462ded5](https://github.com/AElfProject/aelf-block-scan/commit/462ded5))



### [0.3.1](https://github.com/AElfProject/aelf-block-scan/compare/v0.3.0...v0.3.1) (2020-02-08)


### Bug Fixes

* 🐛 fix range when LIB height is larger than  best height ([97cf093](https://github.com/AElfProject/aelf-block-scan/commit/97cf093))



## [0.3.0](https://github.com/AElfProject/aelf-block-scan/compare/v0.2.4...v0.3.0) (2020-02-08)


### Build System

* **deps:** bump eslint-utils from 1.4.0 to 1.4.3 ([c39ce2a](https://github.com/AElfProject/aelf-block-scan/commit/c39ce2a))
* **deps:** bump handlebars from 4.1.2 to 4.7.2 ([ede082c](https://github.com/AElfProject/aelf-block-scan/commit/ede082c))


### Features

* 🎸 use cache when querying unconfirmed blocks ([49e186c](https://github.com/AElfProject/aelf-block-scan/commit/49e186c))



### [0.2.4](https://github.com/AElfProject/aelf-block-scan/compare/v0.2.3...v0.2.4) (2019-12-04)


### Features

* 🎸 do not get tx id ([912275e](https://github.com/AElfProject/aelf-block-scan/commit/912275e))



### [0.2.3](https://github.com/AElfProject/aelf-block-scan/compare/v0.2.2...v0.2.3) (2019-11-18)


### Bug Fixes

* 🐛 it is millisecond!!! ([e04311f](https://github.com/AElfProject/aelf-block-scan/commit/e04311f))



### [0.2.2](https://github.com/AElfProject/aelf-block-scan/compare/v0.2.1...v0.2.2) (2019-11-18)


### Bug Fixes

* 🐛 jump when increasing heights slow ([005830a](https://github.com/AElfProject/aelf-block-scan/commit/005830a))



### [0.2.1](https://github.com/AElfProject/aelf-block-scan/compare/v0.2.0...v0.2.1) (2019-11-18)


### Bug Fixes

* 🐛 remove time log and add case for stop increasing height ([c004cb0](https://github.com/AElfProject/aelf-block-scan/commit/c004cb0))



## [0.2.0](https://github.com/AElfProject/aelf-block-scan/compare/v0.1.6...v0.2.0) (2019-11-18)


### Features

* 🎸 remove useless log ([c1181b3](https://github.com/AElfProject/aelf-block-scan/commit/c1181b3))



### [0.1.6](https://github.com/AElfProject/aelf-block-scan/compare/v0.1.5...v0.1.6) (2019-09-27)


### Bug Fixes

* 🐛 remove console time ([209a177](https://github.com/AElfProject/aelf-block-scan/commit/209a177))



### [0.1.5](https://github.com/AElfProject/aelf-block-scan/compare/v0.1.4...v0.1.5) (2019-08-01)


### Bug Fixes

* 🐛 fix scheduler constructor ([7b2f95b](https://github.com/AElfProject/aelf-block-scan/commit/7b2f95b))



### [0.1.4](https://github.com/AElfProject/aelf-block-scan/compare/v0.1.3...v0.1.4) (2019-08-01)


### Features

* 🎸 export Scheduler ([c18de31](https://github.com/AElfProject/aelf-block-scan/commit/c18de31))



### [0.1.3](https://github.com/AElfProject/aelf-block-scan/compare/v0.1.2...v0.1.3) (2019-07-31)



### [0.1.2](https://github.com/AElfProject/aelf-block-scan/compare/v0.1.1...v0.1.2) (2019-07-31)



### [0.1.1](https://github.com/AElfProject/aelf-block-scan/compare/v0.1.0...v0.1.1) (2019-07-27)



## [0.1.0](https://github.com/AElfProject/aelf-block-scan/compare/v0.7.2-alpha.1...v0.1.0) (2019-07-27)


### refactor

* 💡 refactor scanner ([8881425](https://github.com/AElfProject/aelf-block-scan/commit/8881425))


### BREAKING CHANGES

* remove sql operations, remove formatters


app.js
/*
 * huangzongzhe
 * just a demo, please use scan.js
 */

let config = require('./config.js');
const mysql = require('mysql');
const fs = require('fs');

const rds = require('ali-rds');

const Aelf = require('aelf-sdk');
// const moment = require('moment');

// 我的mbp, egg和node都在本机，200毫秒上下
console.log(JSON.stringify(config));

let mysqlQuery = function(connection, sql, values) {
	return new Promise((resolve, reject) => {
		connection.query(sql, values, (err, rows) => {
			if (err) {
				// connection.rollback();
				reject(err);
			} else {
				resolve(rows);
			}
		});
	});
};

function blockInfoFormat(blockInfo) {
    // {
    //   "Blockhash": "0xab77f8040c5ca21f040d50f7980150137f99421f009b3c6ee4ee6b612d3c20dc",
    //   "Header": {
    //     "PreviousBlockHash": "0x59dfda9c8df722603ff0b0fbda3dd3eaf9a24062dbaed24f63aaa476d7a01a06",
    //     "MerkleTreeRootOfTransactions": "0x554ecbc61102471f65bcc32550f7fddddba7cb90a8f4c3620e79593d7795a942",
    //     "MerkleTreeRootOfWorldState": "0x085df4ecb2b50394fb09a9993828944de0b1cd18e2e55c0c5c3c9b42b1072ea9",
    //     "Index": "1",
    //     "Time": "2018-08-10T03:39:37.625647Z",
    //     "ChainId": "0xaf5de66168d27554dc5baaea887dc373db322921c3965b173eaad0237bc14aca"
    //   },
    //   "Body": {
    //     "TransactionsCount": 1,
    //     "Transactions": [
    //     "0x554ecbc61102471f65bcc32550f7fddddba7cb90a8f4c3620e79593d7795a942"
    //   ]
    //   },
    //   "CurrentTransactionPoolSize": 1
    // }
    // block_hash, pre_block_hash, chain_id, block_height, tx_count, merkle_root_tx, merkle_root_state, time
    let header = blockInfo.Header;
    let body = blockInfo.Body;
    let output = {
        block_hash: blockInfo.Blockhash,
        pre_block_hash: header.PreviousBlockHash,
        chain_id: header.ChainId,
        block_height: header.Index,
        tx_count: body.TransactionsCount,
        merkle_root_tx: header.MerkleTreeRootOfTransactions,
        merkle_root_state: header.MerkleTreeRootOfWorldState,
        time: header.Time // format 2018-08-03 17:26:03
    };
    return output;
}


function transactionFormat(transaction, blockInfo) {
    let method = transaction.tx_info.Method;
    let txInfo = transaction.tx_info;
    // {
    //   "tx_status": "Mined",
    //   "tx_info": {
    //   "TxId": "0x554ecbc61102471f65bcc32550f7fddddba7cb90a8f4c3620e79593d7795a942",
    //   "From": "0x04beb6b3aae09d4ae3b457e9a3066e08519f",
    //   "To": "0x52b297074b3fbcd9be2b9feed931b95917bc",
    //   "Method": "InitializeAElfDPoS",
    //   "IncrementId": 0,
    //   "params": "0x0a24303462656236623361616530396434616533623435376539613330363665303835313966, 0x0a5f0a5d0a243034626562366233616165303964346165336234353765396133303636653038353139661235080110012a220a2051d9bab266f5b03de734f4dbc4a9c91370d958b3b5cfea2d17af34d7736ffbf4320b08fa91b4db0510b8f5a05a0a3b0a390a24303462656236623361616530396434616533623435376539613330363665303835313966121108011001320b088292b4db0510b0ec935b, 0x08a01f"
    //   },
    //   "return": "0x"
    // }
    // 字段搞不清命名的话，去这个规范瞅。
    // http://www.szse.cn/main/files/2015/09/01/深圳证券交易所STEP交易数据接口规范1.00.pdf
    // tx_id, params_to, chain_id, block_height, address_from, address_to, params, method, block_hash, increment_id, quantity(qty)
    let params = txInfo.params.split(', ');
    // 一定要按数据库里字段的顺序排，后面的insertTransactions按这个顺序拼接的数据
    let output = {
        tx_id: txInfo.TxId,
        params_to: '',
        chain_id: blockInfo.chain_id,
        block_height: blockInfo.block_height,
        address_from: txInfo.From,
        address_to: txInfo.To,
        params: txInfo.params,
        method: txInfo.Method,
        block_hash: blockInfo.block_hash,
        increment_id: txInfo.IncrementId,
        quantity: 0
    };
    switch(method) {
        case 'Initialize':
            // 查支出时，需要排除掉method=Initialize这种情况。
            output.params_to = address_from;
            output.quantity = params[2];
            break;
        case 'Transfer':
            output.params_to = params[0];
            output.quantity = params[1];
            break
        default: ;
    }
    return output;
}


function contractFormat(transaction) {
    // contract_address, tx_id, symbol, name, total_supply, decimals
    if (transaction.Method === 'Initialize') {
        let params = transaction.params.split(', ');
        return {
            contract_address: transaction.To,
            tx_id: transaction.TxId,
            symbol: params[0],
            name: params[1],
            total_supply: params[2],
            decimals: params[3]
        }
    }
    return false;
}

/**
 * 批量插入交易
 *
 * @method insertTransactions
 * @param {Array} transactions
 * @param {Object} database
 * @param {String} table
 */
// insertTransactions(txs, aelf0, 'transactions_0');
function insertTransactions(transactions, database, table) {
    // 将max_allowed_packet调大，假设平均500B一条的话，能一次插1.6W条交易。
    // 如果数据实在过大, 拆分数据后再合并数据，多合几条sql。
    // let table = 'transactions_0';
    // let table = table;

    let keys = [
        'tx_id',
        'params_to',
        'chain_id',
        'block_height',
        'address_from',
        'address_to',
        'params',
        'method',
        'block_hash',
        'increment_id',
        'quantity'
    ];
    let valuesBlank = keys.map(() => {
        return '?';
    });

    let keysStr = `(${keys.join(',')})`;
    let values = [];
    let valuesStr = [];

    let output = [];

    transactions.map(item => {
        for (let each in item) {
            values.push(item[each]);
        }

        if (item.method === 'Initialize') {
        	output = insertContract(item, database, 'contract_aelf20');
            // insertContract(item, database, 'contract_aelf20');
        }

        valuesStr.push(`(${valuesBlank.join(',')})`);
    });
    valuesStr = valuesStr.join(',');
    // console.log('sh:::::::::: ', `insert into ${table} ${keysStr} VALUES ${valuesStr};`);
    // console.log('values:: ', values);
	output.push({
		database: database,
    	sql: `insert into ${table} ${keysStr} VALUES ${valuesStr} ON DUPLICATE KEY UPDATE tx_id=VALUES(tx_id);`,
    	values: values
    });
    return output;

    // return mysqlQuery(database, `insert into ${table} ${keysStr} VALUES ${valuesStr} ON DUPLICATE KEY UPDATE tx_id=VALUES(tx_id);`, values);
    // database.query(`insert into ${table} ${keysStr} VALUES ${valuesStr} ON DUPLICATE KEY UPDATE tx_id=VALUES(tx_id);`, values, function (err) {
    //     console.log('insertTransactions err: ', err);
    // });
}

/**
 * 插入区块信息
 *
 * @method blockInfoFormatted
 * @param {Object} blockInfoFormatted
 * @param {Object} database
 * @param {String} table
 */
// insertBlock(blockInfoFormatted, aelf0, 'blocks_0');
function insertBlock(blockInfoFormatted, database, table) {
    let keys = [
        'block_hash',
        'pre_block_hash',
        'chain_id',
        'block_height',
        'tx_count',
        'merkle_root_tx',
        'merkle_root_state',
        'time'
    ];
    let valuesBlank = keys.map(() => {
        return '?';
    });

    let keysStr = `(${keys.join(',')})`;
    let valuesBlankStr = `(${valuesBlank.join(',')})`;

    let values = [];
    for (let each in blockInfoFormatted) {
        values.push(blockInfoFormatted[each]);
    }

    return [
    	{
    		database: database,
    		sql: `insert into ${table} ${keysStr} VALUES ${valuesBlankStr} ON DUPLICATE KEY UPDATE block_hash=VALUES(block_hash);`,
    		values: values
    	}
    ]

    // return mysqlQuery(database, `insert into ${table} ${keysStr} VALUES ${valuesBlankStr} ON DUPLICATE KEY UPDATE block_hash=VALUES(block_hash);`, values);

    // database.query(`insert into ${table} ${keysStr} VALUES ${valuesBlankStr} ON DUPLICATE KEY UPDATE block_hash=VALUES(block_hash);`, values, function (err) {
    //     console.log('insertBlock err:', err);
    // });
}

function insertContract(transaction, database, table) {
    // transaction
    // let keys = [
    //     'tx_id',
    //     'params_to',
    //     'chain_id',
    //     'block_height',
    //     'address_from',
    //     'address_to',
    //     'params',
    //     'method',
    //     'block_hash',
    //     'increment_id',
    //     'quantity'
    // ];
    let keys = [
        'contract_address',
        'tx_id',
        'symbol',
        'name',
        'total_supply',
        'decimals'
    ];
    let valuesBlank = keys.map(() => {
        return '?';
    });

    let keysStr = `(${keys.join(',')})`;
    let valuesBlankStr = `(${valuesBlank.join(',')})`;

    let params = transaction.params.split(', ');
    let values = [transaction['address_to'], transaction['tx_id']].concat(params);

    return [
    	{
    		database: database,
    		sql: `insert into ${table} ${keysStr} VALUES ${valuesBlankStr} ON DUPLICATE KEY UPDATE contract_address=VALUES(contract_address);`,
    		values: values
    	}
    ];
    // return mysqlQuery(database, `insert into ${table} ${keysStr} VALUES ${valuesBlankStr} ON DUPLICATE KEY UPDATE contract_address=VALUES(contract_address);`, values);
    // database.query(`insert into ${table} ${keysStr} VALUES ${valuesBlankStr} ON DUPLICATE KEY UPDATE contract_address=VALUES(contract_address);`, values, function (err) {
    //     console.log('insertContract error: ', err);
    // });
}

let aelf = new Aelf(new Aelf.providers.HttpProvider("http://localhost:1234/chain"));
aelf.chain.connectChain();
let wallet = Aelf.wallet.getWalletByPrivateKey('f6e512a3c259e5f9af981d7f99d245aa5bc52fe448495e0b0dd56e8406be6f71');
var aelf0 = rds(config.mysql.aelf0);
async function scanBlock() {
    // let tokenc = aelf.chain.contractAt('0x358d01d001cd97775e7b7f32fd03f7f28c0a', wallet);

    // console.log(aelf.chain);

	let before = new Date().getTime();

	// 我的mbp, egg和node都在本机，200毫秒上下

	// let result = '';
	// for (let i = 0; i< 100; i++) {
	aelf0.query('select block_height from transactions_0 ORDER BY block_height DESC limit 1', [], async function (error, result, fields) {
		// }
		let blockHeightInDataBase = 0;
		if (result && result[0] && result[0].block_height) {
		    blockHeightInDataBase = parseInt(result[0].block_height, 10);
		}
		let blockHeightInChain = parseInt(aelf.chain.getBlockHeight().result.block_height, 10);
		let blockHeight = 0;

		let beforeGetInfo = new Date().getTime();

		let blockInfo = aelf.chain.getBlockInfo(blockHeight, true);

		// let transactions = blockInfo.result.Body.Transactions;
		let beforeGetTx = new Date().getTime();
		// aelf.chain.getTxResult('0xe31f8a905725456f083996960ca16de1217da2556c31137bd4c15b2f8aa2d1a5', (err, result) => {
		//   console.log('异步1: ', result);
		// });
		// aelf.chain.getTxResult('0xe31f8a905725456f083996960ca16de1217da2556c31137bd4c15b2f8aa2d1a5', (err, result) => {
		//   console.log('异步2: ', result);
		// });
		// aelf.chain.getTxResult('0xe31f8a905725456f083996960ca16de1217da2556c31137bd4c15b2f8aa2d1a5', (err, result) => {
		//   console.log('异步3: ', result);
		// });
		// aelf.chain.getTxResult('0xe31f8a905725456f083996960ca16de1217da2556c31137bd4c15b2f8aa2d1a5', (err, result) => {
		//   console.log('异步4: ', result);
		// });

		// insert into blocks_0 (block_hash,pre_block_hash,chain_id,block_height,tx_count,merkle_root_tx,merkle_root_state,time) VALUES ('test011','223','1',1,2333,'mrt','mrs','2018-08-05 14:20:54') ON DUPLICATE KEY UPDATE block_hash=VALUES(block_hash);
		let errorList = [];
		// let transactionsTemp = [];
		let limit = 100;
		// let max = blockHeightInDataBase + limit;
		let max = 18;
		console.log('blockHeightInDataBase, max', blockHeightInDataBase, ' || ', blockHeightInChain);
		if (blockHeightInDataBase >= blockHeightInChain) {
		    return;
		}

		// N1 ~ N2 个一组，可配置。 默认值100个。
		// 用定时器来检查，如果一个请求和入库都处理完了，请求下一批。
		// 其中抛出异常的区块，全部都记录在案，下一次再处理。
		// 异常测试： output.params_to = output.address_from;  通过删除右侧的output.

		for (let i = blockHeightInDataBase; i <= max; i++) {
		    // 批量查询有了之后，批量拿到信息，然后合并数据走事务插入。
		    // aelf.chain.getBlockInfo(i, true, async (err, result) => {
		    aelf.chain.getBlockInfo(i, true, async (err, result) => {
		        // let tran = await aelf0.beginTransaction(error => {
		        aelf0.beginTransaction(async (error) => {

		            // console.log('beginTransaction: ', error);

			        // try {
			            let blockInfo = result.result;

			            // let insertBlocks0 = 'insert into blocks_0 (block_hash,pre_block_hash,chain_id,block_height,tx_count,merkle_root_tx,merkle_root_state,time) VALUES ('test011','223','1',1,2333,'mrt','mrs','2018-08-05 14:20:54') ON DUPLICATE KEY UPDATE block_hash=VALUES(block_hash);'

			            let transactions = blockInfo.Body.Transactions;

			            let blockInfoFormatted  = blockInfoFormat(blockInfo);

			            console.log('block_hash: ', blockInfoFormatted.block_hash, ' || ', i);

			            let transactionsDetail = [];
			            let transactionsSql = [];
			            // console.log('transactions.length', transactions.length);
			            if (transactions.length) {
			                for (let i = 0, j = transactions.length; i < j; i++) {
			                    // transactionsDetail.push(await aelf.chain.getTxResult(transactions[i]));
			                    let transaction = await aelf.chain.getTxResult(transactions[i]).result;
			                    transaction = transactionFormat(transaction, blockInfoFormatted);
			                    transactionsDetail.push(transaction);
			                }
			                transactionsSql = insertTransactions(transactionsDetail, aelf0, 'transactions_0');
			            }

			            let blocksSql = insertBlock(blockInfoFormatted, aelf0, 'blocks_0');
			            let sql = transactionsSql.concat(blocksSql);

			            const promises = sql.map(item => {
			            	return mysqlQuery(item.database, item.sql, item.values);
			            });

			            Promise.all(promises).then(reuslt => {

			            	aelf0.commit();

			            }).catch(err => {
							aelf0.rollback();

							let file = "./error.json";
							let result = JSON.parse(fs.readFileSync(file));
							result = result || {
								list: []
							};

							console.log('[error]rollback: ', result, i, err);

							result.list.push(i);
							fs.writeFileSync(file, JSON.stringify(result));
			            })

			            // console.log('output: ', output);
			            // console.log(output1.concat(output2));

			            // await tran.commit();
			            // console.log(`transactions${i}: `, transactionsDetail, blockInfoFormatted);
			        // } catch (e) {
			        //     aelf0.rollback();
			        //     errorList.push(blockHeightInDataBase);
			        //     // 错误日志记录
			        //     console.log('getBlockInfo::: ', e, errorList);
			        // }
		        });
		    });
		    // let transactions = blockInfo.result.Body.Transactions;
		    // transactionsTemp.push(transactions);
		}
		// console.log('errrrrrrrrrrrorrrrrr::: ', errorList);

		let after = new Date().getTime();

		// 问：如何保证一个块中的所有的交易都插到了库里。
		let tx_id = Math.random();
    });
}

scanBlock();
// setInterval(function () {
//     scanBlock();
// }, 5000);
